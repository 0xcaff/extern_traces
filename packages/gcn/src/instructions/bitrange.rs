pub struct BitRange {
    pub start: u8,
    pub len: u8,
}

impl BitRange {
    pub fn of_32(&self, value: u32) -> usize {
        let offset = 32 - self.len - self.start;

        let mask = ((1 << self.len) - 1) << offset;

        ((mask & value) >> offset) as usize
    }

    pub fn of_64(&self, value: u64) -> usize {
        let offset = 64 - self.len - self.start;

        let mask = ((1 << self.len) - 1) << offset;

        ((mask & value) >> offset) as usize
    }
}

pub const fn bitrange(start_idx: u8, len: u8) -> BitRange {
    BitRange {
        start: start_idx,
        len,
    }
}
