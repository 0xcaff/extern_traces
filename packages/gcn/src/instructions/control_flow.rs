use crate::instructions::operands::ScalarSourceOperand;
use crate::{
    FormattedInstruction, Instruction, SOP1Instruction, SOP1OpCode, SOP2Instruction, SOP2OpCode,
    SOPKInstruction, SOPKOpCode, SOPPInstruction, SOPPOpCode,
};

#[derive(Debug)]
pub struct ControlFlowInformation {
    pub has_next_instruction: bool,
    pub branch_target: Option<BranchTarget>,
}

#[derive(Debug)]
pub enum BranchTarget {
    Indirect(IndirectValue),
    Direct(u64),
}

#[derive(Debug)]
pub enum IndirectValue {
    Register(ScalarSourceOperand),
}

impl Instruction {
    pub fn control_flow_information(&self) -> ControlFlowInformation {
        match &self.inner {
            FormattedInstruction::SOPP(SOPPInstruction { op, simm16 }) => {
                let branch_target = Some(BranchTarget::Direct(
                    self.program_counter + (*simm16 as u64 * 4) + 4,
                ));

                match op {
                    SOPPOpCode::s_endpgm => {
                        return ControlFlowInformation {
                            has_next_instruction: false,
                            branch_target: None,
                        }
                    }
                    SOPPOpCode::s_branch => {
                        return ControlFlowInformation {
                            branch_target,
                            has_next_instruction: false,
                        }
                    }
                    SOPPOpCode::s_cbranch_vccz
                    | SOPPOpCode::s_cbranch_vccnz
                    | SOPPOpCode::s_cbranch_execz
                    | SOPPOpCode::s_cbranch_execnz
                    | SOPPOpCode::s_cbranch_scc0
                    | SOPPOpCode::s_cbranch_scc1
                    | SOPPOpCode::s_cbranch_cdbgsys
                    | SOPPOpCode::s_cbranch_cdbguser
                    | SOPPOpCode::s_cbranch_cdbgsys_and_user
                    | SOPPOpCode::s_cbranch_cdbgsys_or_user => {
                        return ControlFlowInformation {
                            branch_target,
                            has_next_instruction: true,
                        }
                    }
                    SOPPOpCode::s_trap => {
                        unimplemented!("trap handler not implemented")
                    }
                    _ => {}
                }
            }
            FormattedInstruction::SOP1(SOP1Instruction { op, sdst, ssrc0 }) => match op {
                SOP1OpCode::s_setpc_b64 | SOP1OpCode::s_swappc_b64 | SOP1OpCode::s_rfe_b64 => {
                    return ControlFlowInformation {
                        has_next_instruction: false,
                        branch_target: Some(BranchTarget::Indirect(IndirectValue::Register(
                            ssrc0.clone(),
                        ))),
                    }
                }
                _ => {}
            },
            FormattedInstruction::SOP2(SOP2Instruction {
                op: SOP2OpCode::s_cbranch_g_fork,
                ssrc1,
                ..
            }) => {
                return ControlFlowInformation {
                    has_next_instruction: true,
                    branch_target: Some(BranchTarget::Indirect(IndirectValue::Register(
                        ssrc1.clone(),
                    ))),
                }
            }
            FormattedInstruction::SOPK(SOPKInstruction {
                op: SOPKOpCode::s_cbranch_i_fork,
                simm16,
                ..
            }) => {
                return ControlFlowInformation {
                    has_next_instruction: false,
                    branch_target: Some(BranchTarget::Direct(
                        self.program_counter + 4 + (*simm16 as u64),
                    )),
                }
            }
            _ => {}
        };

        ControlFlowInformation {
            has_next_instruction: true,
            branch_target: None,
        }
    }
}
