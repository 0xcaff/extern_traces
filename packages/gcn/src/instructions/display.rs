pub trait DisplayInstruction {
    fn display(&self) -> DisplayableInstruction;
}

#[derive(Debug)]
pub struct DisplayableInstruction {
    pub op: String,
    pub args: Vec<String>,
}
