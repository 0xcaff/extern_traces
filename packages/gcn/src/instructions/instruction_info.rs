pub struct InstructionInfo {
    pub definitions: &'static [Option<OperandInfo>; 4],
    pub operands: &'static [Option<OperandInfo>; 4],
}

pub enum OperandInfo {
    M0,
    SCC,
    Exec,
    ExecLo,
    Vcc,
    Size(u8),
}
