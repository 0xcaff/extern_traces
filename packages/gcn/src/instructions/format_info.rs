use crate::instructions::{OpFormat, OpInfo, OPS};
#[macro_export]
macro_rules! op_format_pattern {
    ($bitstring:expr) => {{
        const WIDTH: usize = $bitstring.len();
        const BITS: u32 = {
            let bytes = $bitstring.as_bytes();
            let mut bits = 0;

            let mut i = 0;
            while i < WIDTH {
                bits <<= 1;
                match bytes[i] {
                    b'0' => {}
                    b'1' => bits |= 1,
                    _ => panic!("Invalid character in bitstring"), // Causes compile-time error
                }
                i += 1;
            }
            bits
        };

        OpFormatPattern {
            width: WIDTH as u8,
            bits: BITS,
        }
    }};
}

pub const OP_FORMATS: [OpFormatInfo; 25] = [
    OpFormatInfo {
        format: OpFormat::SOP1,
        bytes_len: 4,
        pattern: Some(op_format_pattern!("101111101")),
        op_bits: Some(bitrange(16, 8)),
    },
    OpFormatInfo {
        format: OpFormat::SOP2,
        bytes_len: 4,
        pattern: Some(op_format_pattern!("10")),
        op_bits: Some(bitrange(2, 7)),
    },
    OpFormatInfo {
        format: OpFormat::SOPK,
        bytes_len: 4,
        pattern: Some(op_format_pattern!("1011")),
        op_bits: Some(bitrange(4, 5)),
    },
    OpFormatInfo {
        format: OpFormat::SOPP,
        bytes_len: 4,
        pattern: Some(op_format_pattern!("101111111")),
        op_bits: Some(bitrange(9, 7)),
    },
    OpFormatInfo {
        format: OpFormat::SOPC,
        bytes_len: 4,
        pattern: Some(op_format_pattern!("101111100")),
        op_bits: Some(bitrange(9, 7)),
    },
    OpFormatInfo {
        format: OpFormat::SMEM,
        bytes_len: 4,
        pattern: Some(op_format_pattern!("11000")),
        op_bits: None,
    },
    OpFormatInfo {
        format: OpFormat::DS,
        bytes_len: 8,
        pattern: Some(op_format_pattern!("110110")),
        op_bits: Some(bitrange(6, 8)),
    },
    OpFormatInfo {
        format: OpFormat::LDSDIR,
        bytes_len: 8,
        pattern: None,
        op_bits: None,
    },
    OpFormatInfo {
        format: OpFormat::MTBUF,
        bytes_len: 8,
        pattern: Some(op_format_pattern!("111010")),
        op_bits: Some(bitrange(13, 3)),
    },
    OpFormatInfo {
        format: OpFormat::MUBUF,
        bytes_len: 8,
        pattern: Some(op_format_pattern!("111000")),
        op_bits: Some(bitrange(7, 7)),
    },
    OpFormatInfo {
        format: OpFormat::MIMG,
        bytes_len: 8,
        pattern: Some(op_format_pattern!("111100")),
        op_bits: Some(bitrange(7, 7)),
    },
    OpFormatInfo {
        format: OpFormat::EXP,
        bytes_len: 8,
        pattern: Some(op_format_pattern!("111110")),
        op_bits: None,
    },
    OpFormatInfo {
        format: OpFormat::FLAT,
        bytes_len: 8,
        pattern: None,
        op_bits: None,
    },
    OpFormatInfo {
        format: OpFormat::GLOBAL,
        bytes_len: 8,
        pattern: None,
        op_bits: None,
    },
    OpFormatInfo {
        format: OpFormat::SCRATCH,
        bytes_len: 8,
        pattern: None,
        op_bits: None,
    },
    OpFormatInfo {
        format: OpFormat::VINTRP,
        bytes_len: 4,
        pattern: Some(op_format_pattern!("110010")),
        op_bits: Some(bitrange(14, 2)),
    },
    OpFormatInfo {
        format: OpFormat::VINTERP_INREG,
        bytes_len: 8,
        pattern: None,
        op_bits: None,
    },
    OpFormatInfo {
        format: OpFormat::VOP1,
        bytes_len: 4,
        pattern: Some(op_format_pattern!("0111111")),
        op_bits: Some(bitrange(15, 8)),
    },
    OpFormatInfo {
        format: OpFormat::VOP2,
        bytes_len: 4,
        pattern: Some(op_format_pattern!("0")),
        op_bits: Some(bitrange(1, 6)),
    },
    OpFormatInfo {
        format: OpFormat::VOPC,
        bytes_len: 4,
        pattern: Some(op_format_pattern!("0111110")),
        op_bits: Some(bitrange(7, 8)),
    },
    OpFormatInfo {
        format: OpFormat::VOP3,
        bytes_len: 8,
        pattern: Some(op_format_pattern!("110100")),
        op_bits: Some(bitrange(6, 9)),
    },
    OpFormatInfo {
        format: OpFormat::VOP3P,
        bytes_len: 8,
        pattern: None,
        op_bits: None,
    },
    OpFormatInfo {
        format: OpFormat::SDWA,
        bytes_len: 4,
        pattern: None,
        op_bits: None,
    },
    OpFormatInfo {
        format: OpFormat::DPP16,
        bytes_len: 4,
        pattern: None,
        op_bits: None,
    },
    OpFormatInfo {
        format: OpFormat::DPP8,
        bytes_len: 8,
        pattern: None,
        op_bits: None,
    },
];

pub struct OpFormatInfo {
    pub format: OpFormat,
    pub bytes_len: u8,

    pub pattern: Option<OpFormatPattern>,
    pub op_bits: Option<BitRange>,
}

pub struct OpFormatPattern {
    pub width: u8,
    pub bits: u32,
}

pub struct BitRange {
    pub start: u8,
    pub len: u8,
}

impl BitRange {
    pub fn of(&self, value: u32) -> u32 {
        let offset = 32 - self.len - self.start;

        let mask = ((1 << self.len) - 1) << offset;

        (mask & value) >> offset
    }
}

pub const fn bitrange(start: u8, len: u8) -> BitRange {
    BitRange { start, len }
}


impl OpFormat {
    pub fn ops(&self) -> impl Iterator<Item = &'static OpInfo> + '_ {
        OPS.iter().filter(move |it| &it.format == self)
    }

    pub fn info(&self) -> &'static OpFormatInfo {
        &OP_FORMATS[(*self) as usize]
    }
}
