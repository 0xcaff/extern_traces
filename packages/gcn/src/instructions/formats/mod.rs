pub use crate::instructions::formats::ds::DSInstruction;
pub use crate::instructions::formats::exp::ExpInstruction;
pub use crate::instructions::formats::mimg::MIMGInstruction;
pub use crate::instructions::formats::mtbuf::MTBufInstruction;
pub use crate::instructions::formats::mubuf::MUBUFInstruction;
pub use crate::instructions::formats::smem::SMEMInstruction;
pub use crate::instructions::formats::sop1::SOP1Instruction;
pub use crate::instructions::formats::sop2::SOP2Instruction;
pub use crate::instructions::formats::sopc::SOPCInstruction;
pub use crate::instructions::formats::sopk::SOPKInstruction;
pub use crate::instructions::formats::sopp::SOPPInstruction;
pub use crate::instructions::formats::vintrp::VINTRPInstruction;
pub use crate::instructions::formats::vop1::VOP1Instruction;
pub use crate::instructions::formats::vop2::VOP2Instruction;
pub use crate::instructions::formats::vop3::VOP3Instruction;
pub use crate::instructions::formats::vopc::VOPCInstruction;
use crate::instructions::operands::{ScalarSourceOperand, SourceOperand};
use crate::instructions::ops::{SOPKOpCode, VOP1OpCode, VOP2OpCode};
use crate::reader::EofError;
use crate::SliceReader;
use bits::FromBits;
use gcn_internal_macros::{DisplayInstruction, ParseInstruction};

pub mod ds;
pub mod exp;
pub mod mimg;
pub mod mtbuf;
pub mod mubuf;
pub mod smem;
pub mod sop1;
pub mod sop2;
pub mod sopc;
pub mod sopk;
pub mod sopp;
pub mod vintrp;
pub mod vop1;
pub mod vop2;
pub mod vop3;
pub mod vopc;

#[derive(Debug, ParseInstruction, DisplayInstruction)]
pub enum FormattedInstruction {
    #[pattern(0b110110)]
    DS(DSInstruction),

    #[pattern(0b111110)]
    EXP(ExpInstruction),

    #[pattern(0b111100)]
    MIMG(MIMGInstruction),

    #[pattern(0b111010)]
    MTBUF(MTBufInstruction),

    #[pattern(0b111000)]
    MUBUF(MUBUFInstruction),

    #[pattern(0b11000)]
    SMEM(SMEMInstruction),

    #[pattern(0b101111101)]
    SOP1(SOP1Instruction),

    #[pattern(0b10)]
    SOP2(SOP2Instruction),

    #[pattern(0b101111100)]
    SOPC(SOPCInstruction),

    #[pattern(0b1011)]
    SOPK(SOPKInstruction),

    #[pattern(0b101111111)]
    SOPP(SOPPInstruction),

    #[pattern(0b110010)]
    VINTRP(VINTRPInstruction),

    #[pattern(0b0111111)]
    VOP1(VOP1Instruction),

    #[pattern(0b0)]
    VOP2(VOP2Instruction),

    #[pattern(0b110100)]
    VOP3(VOP3Instruction),

    #[pattern(0b0111110)]
    VOPC(VOPCInstruction),
}

impl FormattedInstruction {
    pub fn has_literal_constant(&self) -> bool {
        matches!(
            self,
            FormattedInstruction::SOP2(
                SOP2Instruction {
                    ssrc0: ScalarSourceOperand::LiteralConstant,
                    ..
                } | SOP2Instruction {
                    ssrc1: ScalarSourceOperand::LiteralConstant,
                    ..
                },
            ) | FormattedInstruction::SOPK(SOPKInstruction {
                op: SOPKOpCode::s_setreg_imm32_b32,
                ..
            }) | FormattedInstruction::SOP1(SOP1Instruction {
                ssrc0: ScalarSourceOperand::LiteralConstant,
                ..
            }) | FormattedInstruction::SOPC(
                SOPCInstruction {
                    ssrc0: ScalarSourceOperand::LiteralConstant,
                    ..
                } | SOPCInstruction {
                    ssrc1: ScalarSourceOperand::LiteralConstant,
                    ..
                },
            ) | FormattedInstruction::VOP1(VOP1Instruction {
                src0: SourceOperand::Scalar(ScalarSourceOperand::LiteralConstant),
                ..
            }) | FormattedInstruction::VOP2(
                VOP2Instruction {
                    src0: SourceOperand::Scalar(ScalarSourceOperand::LiteralConstant),
                    ..
                } | VOP2Instruction {
                    op: VOP2OpCode::v_madmk_f32 | VOP2OpCode::v_madak_f32,
                    ..
                },
            ) | FormattedInstruction::VOPC(VOPCInstruction {
                src0: SourceOperand::Scalar(ScalarSourceOperand::LiteralConstant),
                ..
            })
        )
    }

    /// Whether this instruction is disabled if the thread's exec bit is 0. Instructions like
    /// v_cmp_gt_u32 which are modified by the exec mask but not disabled by it will not be
    /// included.
    pub fn exec_mask_disables(&self) -> bool {
        fn vop3(vop3: &vop3::OpCode) -> bool {
            match vop3 {
                vop3::OpCode::VOP3(_) => {
                    unimplemented!()
                }
                vop3::OpCode::VOPC(..) => false,
                vop3::OpCode::VOP2(op) => match op {
                    VOP2OpCode::v_readlane_b32 | VOP2OpCode::v_writelane_b32 => false,
                    _ => true,
                },
                vop3::OpCode::VOP1(op) => match op {
                    VOP1OpCode::v_readfirstlane_b32 => false,
                    _ => true,
                },
            }
        }

        match self {
            // Vector ALU
            FormattedInstruction::VOP2(VOP2Instruction { op, .. }) => {
                vop3(&vop3::OpCode::VOP2(*op))
            }
            FormattedInstruction::VOP1(VOP1Instruction { op, .. }) => {
                vop3(&vop3::OpCode::VOP1(*op))
            }
            FormattedInstruction::VOP3(VOP3Instruction { op, .. }) => vop3(op),
            FormattedInstruction::VINTRP(..) => true,

            // Vector Memory
            FormattedInstruction::MUBUF(..)
            | FormattedInstruction::MTBUF(..)
            | FormattedInstruction::MIMG(..) => true,

            // Local and Global Data Share
            FormattedInstruction::DS(..) => true,

            // Export
            FormattedInstruction::EXP(..) => true,

            // Scalar Operations (starting with s_)
            FormattedInstruction::SMEM(..)
            | FormattedInstruction::SOP1(..)
            | FormattedInstruction::SOP2(..)
            | FormattedInstruction::SOPC(..)
            | FormattedInstruction::SOPK(..)
            | FormattedInstruction::SOPP(..) => false,

            // VOPC is handled at instruction level
            FormattedInstruction::VOPC(..) => false,
        }
    }
}

pub trait ParseInstruction {
    fn parse(token: u32, reader: &mut SliceReader) -> Result<Self, anyhow::Error>
    where
        Self: Sized;
}

impl<T: FromBits<32>> ParseInstruction for T {
    fn parse(token: u32, _reader: &mut SliceReader) -> Result<Self, anyhow::Error>
    where
        Self: Sized,
    {
        Ok(Self::from_bits(token))
    }
}

fn combine(first_token: u32, reader: &mut SliceReader) -> Result<u64, EofError> {
    let second_token = reader.read_u32()?;
    let token = (first_token as u64) | ((second_token as u64) << 32);

    Ok(token)
}
