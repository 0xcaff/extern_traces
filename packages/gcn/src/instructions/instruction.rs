use crate::instructions::OpCode;

pub struct Instruction {
    pub code: OpCode,
}
