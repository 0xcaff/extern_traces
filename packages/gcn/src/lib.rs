#![feature(buf_read_has_data_left)]

mod instructions;
mod reader;

pub use instructions::Instruction;

#[cfg(test)]
mod tests {
    use crate::instructions::Decoder;
    use std::io::Cursor;

    #[test]
    fn test_cube_vertex_shader() {
        let data: [u8; 0x78] = [
            0x00, 0x21, 0x80, 0xBE, 0x00, 0x07, 0x80, 0xC0, 0x7F, 0xC0, 0x8C, 0xBF, 0x08, 0x01,
            0x82, 0xC2, 0x0C, 0x01, 0x84, 0xC2, 0x04, 0x01, 0x86, 0xC2, 0x00, 0x01, 0x80, 0xC2,
            0x71, 0x00, 0x8C, 0xBF, 0x04, 0x0C, 0x0E, 0x10, 0x05, 0x0C, 0x00, 0x10, 0x06, 0x0C,
            0x02, 0x10, 0x07, 0x0C, 0x0C, 0x10, 0x08, 0x0E, 0x0E, 0x06, 0x09, 0x00, 0x00, 0x06,
            0x0A, 0x02, 0x02, 0x06, 0x0B, 0x0C, 0x0C, 0x06, 0x0C, 0x0A, 0x0E, 0x3E, 0x0D, 0x0A,
            0x00, 0x3E, 0x0E, 0x0A, 0x02, 0x3E, 0x0F, 0x0A, 0x0C, 0x3E, 0x00, 0x08, 0x0E, 0x3E,
            0x01, 0x08, 0x00, 0x3E, 0x02, 0x08, 0x02, 0x3E, 0x03, 0x08, 0x0C, 0x3E, 0xCF, 0x08,
            0x00, 0xF8, 0x07, 0x00, 0x01, 0x06, 0x70, 0x3F, 0x8C, 0xBF, 0x03, 0x02, 0x00, 0xF8,
            0x08, 0x09, 0x80, 0x80, 0x00, 0x00, 0x81, 0xBF,
        ];

        insta::assert_snapshot!(test_shader(&data));
    }

    #[test]
    fn test_cube_fragment_shader() {
        let data: [u8; 0x4C] = [
            0x7E, 0x0A, 0xFE, 0xBE, 0x00, 0x01, 0xC4, 0xC0, 0x08, 0x01, 0x80, 0xC0, 0x04, 0x03,
            0xFC, 0xBE, 0x00, 0x00, 0x08, 0xC8, 0x01, 0x00, 0x09, 0xC8, 0x00, 0x01, 0x00, 0xC8,
            0x01, 0x01, 0x01, 0xC8, 0x00, 0x03, 0x06, 0x7E, 0x7F, 0xC0, 0x8C, 0xBF, 0x00, 0x07,
            0x80, 0xF0, 0x02, 0x00, 0x02, 0x00, 0x70, 0x3F, 0x8C, 0xBF, 0x00, 0x03, 0x00, 0x5E,
            0x01, 0x00, 0x5E, 0xD2, 0x02, 0xE5, 0x01, 0x00, 0x0F, 0x1C, 0x00, 0xF8, 0x00, 0x01,
            0x80, 0x80, 0x00, 0x00, 0x81, 0xBF,
        ];

        insta::assert_snapshot!(test_shader(&data));
    }

    fn test_shader(shader_bytes: &[u8]) -> String {
        let cursor = Cursor::new(shader_bytes);
        let mut decoder = Decoder::new(cursor);

        let mut result = vec![];

        loop {
            let instruction = match decoder.decode().unwrap() {
                Some(value) => {
                    result.push(value);
                }
                None => break,
            };
        }

        format!("{:#?}", result)
    }
}
