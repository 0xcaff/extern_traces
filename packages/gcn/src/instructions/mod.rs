mod formats;
mod generated;
mod operands;

use byteorder::{LittleEndian, ReadBytesExt};
use std::io::{BufRead, BufReader, Read};

use crate::instructions::formats::{FormattedInstruction, ParseInstruction};

pub struct Decoder<R> {
    reader: BufReader<R>,
}

#[derive(Debug)]
pub struct Instruction {
    inner: FormattedInstruction,
}

impl<R> Decoder<R>
where
    R: Read,
{
    pub fn new(reader: R) -> Decoder<R> {
        Decoder {
            reader: BufReader::new(reader),
        }
    }

    pub fn decode(&mut self) -> Result<Option<Instruction>, anyhow::Error> {
        // todo: rework this, doesn't handle branching or termination correctly
        if !self.reader.has_data_left()? {
            return Ok(None);
        }

        let token = self.reader.read_u32::<LittleEndian>()?;

        let inner = FormattedInstruction::parse(token, &mut self.reader)?;

        Ok(Some(Instruction { inner }))
    }
}
