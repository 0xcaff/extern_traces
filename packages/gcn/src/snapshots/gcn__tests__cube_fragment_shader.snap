---
source: packages/gcn/src/lib.rs
expression: test_shader(&data)
---
[
    Instruction {
        inner: SOP1(
            SOP1Instruction {
                op: s_wqm_b64,
                ssrc0: Destination(
                    ExecLo,
                ),
                sdst: ExecLo,
            },
        ),
    },
    Instruction {
        inner: SMEM(
            SMEMInstruction {
                op: s_load_dwordx8,
                offset: 0,
                imm: true,
                sbase: 0,
                sdst: ScalarGPR(
                    8,
                ),
            },
        ),
    },
    Instruction {
        inner: SMEM(
            SMEMInstruction {
                op: s_load_dwordx4,
                offset: 8,
                imm: true,
                sbase: 0,
                sdst: ScalarGPR(
                    0,
                ),
            },
        ),
    },
    Instruction {
        inner: SOP1(
            SOP1Instruction {
                op: s_mov_b32,
                ssrc0: Destination(
                    ScalarGPR(
                        4,
                    ),
                ),
                sdst: M0,
            },
        ),
    },
    Instruction {
        inner: VINTRP(
            VINTRPInstruction {
                op: v_interp_p1_f32,
                vsrc: VectorGPR {
                    register_idx: 0,
                },
                attrchan: AttrChan(
                    0,
                ),
                attr: Attr(
                    0,
                ),
                vdst: VectorGPR {
                    register_idx: 2,
                },
            },
        ),
    },
    Instruction {
        inner: VINTRP(
            VINTRPInstruction {
                op: v_interp_p2_f32,
                vsrc: VectorGPR {
                    register_idx: 1,
                },
                attrchan: AttrChan(
                    0,
                ),
                attr: Attr(
                    0,
                ),
                vdst: VectorGPR {
                    register_idx: 2,
                },
            },
        ),
    },
    Instruction {
        inner: VINTRP(
            VINTRPInstruction {
                op: v_interp_p1_f32,
                vsrc: VectorGPR {
                    register_idx: 0,
                },
                attrchan: AttrChan(
                    1,
                ),
                attr: Attr(
                    0,
                ),
                vdst: VectorGPR {
                    register_idx: 0,
                },
            },
        ),
    },
    Instruction {
        inner: VINTRP(
            VINTRPInstruction {
                op: v_interp_p2_f32,
                vsrc: VectorGPR {
                    register_idx: 1,
                },
                attrchan: AttrChan(
                    1,
                ),
                attr: Attr(
                    0,
                ),
                vdst: VectorGPR {
                    register_idx: 0,
                },
            },
        ),
    },
    Instruction {
        inner: VOP1(
            VOP1Instruction {
                op: v_mov_b32,
                src0: VectorGPR(
                    VectorGPR {
                        register_idx: 0,
                    },
                ),
                vdst: VectorGPR {
                    register_idx: 3,
                },
            },
        ),
    },
    Instruction {
        inner: SOPP(
            SOPPInstruction {
                op: s_waitcnt,
                simm16: 49279,
            },
        ),
    },
    Instruction {
        inner: MIMG(
            MIMGInstruction {
                op: image_load,
                dmask: DMask(
                    0,
                ),
                unorm: false,
                glc: false,
                da: false,
                r128: false,
                tfe: false,
                lwe: true,
                slc: false,
                vaddr: VectorGPR {
                    register_idx: 0,
                },
                vdata: VectorGPR {
                    register_idx: 7,
                },
            },
        ),
    },
    Instruction {
        inner: SOPP(
            SOPPInstruction {
                op: s_waitcnt,
                simm16: 16240,
            },
        ),
    },
    Instruction {
        inner: VOP2(
            VOP2Instruction {
                op: v_cvt_pkrtz_f16_f32,
                src0: VectorGPR(
                    VectorGPR {
                        register_idx: 0,
                    },
                ),
                vsrc1: VectorGPR {
                    register_idx: 1,
                },
                vdst: VectorGPR {
                    register_idx: 0,
                },
            },
        ),
    },
    Instruction {
        inner: VOP3(
            VOP3Instruction {
                op: VOPC(
                    v_cmp_f_f32,
                ),
                vdst: VectorGPR {
                    register_idx: 2,
                },
                src0: TransformedOperand {
                    operand: Scalar(
                        Destination(
                            ScalarGPR(
                                1,
                            ),
                        ),
                    ),
                    abs: true,
                    neg: true,
                },
                src1: TransformedOperand {
                    operand: VectorGPR(
                        VectorGPR {
                            register_idx: 0,
                        },
                    ),
                    abs: false,
                    neg: true,
                },
                src2: TransformedOperand {
                    operand: Scalar(
                        IntegerConstant(
                            InlineIntegerConstant {
                                value: 151,
                            },
                        ),
                    ),
                    abs: true,
                    neg: false,
                },
                clamp: false,
            },
        ),
    },
    Instruction {
        inner: EXP(
            ExpInstruction {
                en: En(
                    0,
                ),
                tgt: Tgt(
                    16,
                ),
                compr: false,
                done: false,
                vm: false,
                vsrc0: VectorGPR {
                    register_idx: 15,
                },
                vsrc1: VectorGPR {
                    register_idx: 28,
                },
                vsrc2: VectorGPR {
                    register_idx: 0,
                },
                vsrc3: VectorGPR {
                    register_idx: 248,
                },
            },
        ),
    },
    Instruction {
        inner: SOPP(
            SOPPInstruction {
                op: s_endpgm,
                simm16: 0,
            },
        ),
    },
]
